### PR TITLE
CT-1308 Ability to link cases 

### PR DESCRIPTION
**Pull request by aliuk2012**

### CODE REVIEW

File: app/decorators/case_links_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class CaseLinksDecorator < Draper::Decorator
```
Comment (beccapearce) [inline app/decorators/case_links_decorator.rb]: This decorator is plural while the others are singular

File: app/helpers/cases_helper.rb [hunk 2]
```diff
@@ -136,24 +136,37 @@ def case_uploaded_request_files_id
...
-  def action_link_for_extend_for_pit
+  def action_link_for_extend_for_pit(kase)
     link_to I18n.t('common.case.extend_for_pit'),
-            extend_for_pit_case_path(@case),
+            extend_for_pit_case_path(kase),
             id: 'action--extend-for-pit'
 
   end
 
+  def action_link_for_new_link(kase)
```
Comment (beccapearce) [inline app/helpers/cases_helper.rb]: This is a little ambiguous, maybe action_create_link ? although it's not consistent , not sure what the better option is.

File: app/state_machines/cases/foi_state_machine.rb [hunk 3]
```diff
@@ -288,6 +288,17 @@ def self.get_policy(user_id, object)
     transition from: :awaiting_dispatch,                to: :awaiting_dispatch
   end
 
+  event :linked_a_case do
```
Comment (beccapearce) [inline app/state_machines/cases/foi_state_machine.rb]: should this be a past tense? if so would case_linked be better

File: app/helpers/cases_helper.rb (hunk 2 above)
Comment (aliuk2012) [reply app/helpers/cases_helper.rb]: You are right that is "new_link" is a little ambigous, Unfortunately we will have to keep the `action_link_for` has this method is called from `action_links_for_allowed_events` where it appends the event name to the method.  However what I have done to try address this issue is renamed all the "new_link" to "new_case_link" so its obvious as to what new link it is.

File: app/controllers/cases_controller.rb [hunk 4]
```diff
@@ -390,12 +392,41 @@ def request_further_clearance
...
+
+    @case = CaseLinkDecorator.decorate @case
+  end
+
+  def execute_new_case_link
+    authorize @case, :new_case_link?
+
+    new_case_link_params = params[:case][:linked_case_number]
+
+    service = CaseLinkingService.new current_user, @case, new_case_link_params
```
Comment (misaka) [inline app/controllers/cases_controller.rb]: This initialiser takes `link_case_number` not `new_case_link_params` as the last param. Confusing ... maybe we should rename `new_case_link_params` here to `link_case_number` or something similar.

File: app/services/case_linking_service.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    @link_case = nil
+    @result = :incomplete
+  end
+
+  def call
+    if validate_params
+      ActiveRecord::Base.transaction do
+
+        #Create the links
+        @case.add_linked_case(@link_case)
```
Comment (misaka) [inline app/services/case_linking_service.rb]: If I read correctly, `Case#add_linked_case` returns `:success` or `:failed` ... so if that fails we don't catch that here to reflect that in `@result`, and end up creating the transitions below. Shouldn't our behaviour change?

File: app/services/case_linking_service.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+  private
+
+  def validate_params
+    unless @link_case_number.present?
+      @case.errors.add(:linked_case_number, "can't be blank")
+      @result = :validation_error
+    end
+
+    unless @link_case_number != @case.number
```
Comment (misaka) [inline app/services/case_linking_service.rb]: It's probably just me, but I find this hard to read. Wouldn't `if @link_case_number == @case.number` be the same and also be easier to read?

File: app/controllers/cases_controller.rb (hunk 4 above)
Comment (aliuk2012) [reply app/controllers/cases_controller.rb]: 👍

**Comment by misaka** [comment]
Ok, specs look good, nothing to add.

File: app/policies/cases/base_policy.rb [hunk 7]
```diff
@@ -41,6 +41,15 @@ def request_further_clearance?
           check_can_trigger_event(:request_further_clearance)
     end
 
+    def new_case_link?
+      clear_failed_checks
+      check_user_is_a_manager_for_case ||
+          (!self.case.awaiting_responder? && check_user_is_a_responder_for_case) ||
+          check_user_is_an_approver_for_case
+
```
[the remark this answers is not in the document]
Comment (aliuk2012) [reply app/policies/cases/base_policy.rb]: Good spot.  I think BMT will want to link closed cases for internal reviews so I've added that to the transition event

File: app/services/case_linking_service.rb (hunk 6 above)
Comment (aliuk2012) [reply app/services/case_linking_service.rb]: Its just you  😈  😝 

I've updated the condition

File: app/policies/cases/base_policy.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (misaka) [reply app/policies/cases/base_policy.rb]: Cool. The only thing I'd say is that order of operations with logical operators isn't always obvious ... it'd be nicer to the reader if this expression had parenthesis to be clear to anyone reading (I had to double-check to be sure that it was right):

```
check_can_trigger_event(:link_a_case) &&
    (check_user_is_a_manager_for_case ||
     (!self.case.awaiting_responder? &&
         check_user_is_a_responder_for_case) ||
     check_user_is_an_approver_for_case)
```

I'll still approve, maybe something to keep in mind for next time.

**Pull request merged**